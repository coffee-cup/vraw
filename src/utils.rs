#[derive(Debug, PartialEq, Clone, Copy)]
pub struct Pos {
    pub line: u32,
    pub column: u32,
}

#[derive(Debug, PartialEq, Clone, Copy)]
pub struct Range {
    pub start: Pos,
    pub end: Pos,
}
