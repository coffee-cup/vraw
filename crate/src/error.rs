use std::fmt;

use crate::utils::*;

#[derive(Debug, PartialEq)]
pub struct Error<T: fmt::Display> {
    pub error_type: T,
    pub pos: Pos,
}

impl<T: fmt::Display> Error<T> {
    pub fn new(error_type: T, pos: Pos) -> Error<T> {
        Error {
            error_type: error_type,
            pos: pos,
        }
    }
}

impl<T: fmt::Display> HasPos for Error<T> {
    fn pos(&self) -> Pos {
        self.pos
    }
}

impl<T: fmt::Display> fmt::Display for Error<T> {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(
            f,
            "[{}:{}]: {}",
            self.pos.line, self.pos.column, self.error_type
        )
    }
}
