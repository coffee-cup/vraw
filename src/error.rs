use std::fmt;

use crate::utils::*;

#[derive(Debug, PartialEq)]
pub struct Error {
    message: String,
    pos: Option<Pos>,
}

impl Error {
    pub fn new(message: String, pos: Option<Pos>) -> Error {
        Error {
            message: message,
            pos: pos,
        }
    }
}

impl fmt::Display for Error {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self.pos {
            Some(p) => write!(f, "[{}:{}]: {}", p.line, p.column, self.message),
            _ => write!(f, "{}", self.message),
        }
    }
}
