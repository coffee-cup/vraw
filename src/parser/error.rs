use std::fmt;

use crate::utils::*;

#[derive(Debug, PartialEq)]
pub enum ParseErrorType {
    UnExpectedEndOfInput,
    IdentiferCannotBeReservedWord(String),
    UnBalancedParen,
    Expected(String, Option<String>),
}

#[derive(Debug, PartialEq)]
pub struct ParseError {
    pub error_type: ParseErrorType,
    pub pos: Pos,
}

impl HasPos for ParseError {
    fn pos(&self) -> Pos {
        self.pos
    }
}

impl fmt::Display for ParseErrorType {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            ParseErrorType::UnExpectedEndOfInput => write!(f, "Unexpected end of input."),
            ParseErrorType::IdentiferCannotBeReservedWord(id) => {
                write!(f, "Identifier {} cannot be a reserved word.", id)
            }
            ParseErrorType::UnBalancedParen => write!(f, "Unbalanced paren."),
            ParseErrorType::Expected(expected, found) => match found {
                Some(found) => write!(f, "Expected {}. Found {}.", expected, found),
                None => write!(f, "Expected {}", expected),
            },
        }
    }
}

impl fmt::Display for ParseError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(
            f,
            "[{}:{}] {}",
            self.pos.line, self.pos.column, self.error_type
        )
    }
}
