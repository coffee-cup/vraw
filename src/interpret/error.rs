use std::fmt;

use crate::error::Error;
use crate::utils::*;

#[derive(Debug, PartialEq, Clone)]
pub enum EvalErrorType {
    VariableNotDefined(String),
    ShapeNotDefined(String),
    TypeMismatch(String, String),
    SvgExpectsString(String),
    ShapeAlreadyDefined(String),
    NumArgs(String, usize, usize),
    InvalidArgName(String, String),
    MissingArgs(String, Vec<String>),
    MissingMain,
}

impl fmt::Display for EvalErrorType {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            EvalErrorType::VariableNotDefined(name) => write!(f, "Variable `{}` not defined", name),
            EvalErrorType::ShapeNotDefined(name) => write!(f, "Shape `{}` not defined", name),
            EvalErrorType::TypeMismatch(expected, received) => write!(
                f,
                "Typemismatch. Expected: {}, Received: {}",
                expected, received
            ),
            EvalErrorType::SvgExpectsString(received) => write!(
                f,
                "svg value arg needs to be a string. Received {}",
                received
            ),
            EvalErrorType::ShapeAlreadyDefined(name) => write!(f, "Shape {} already defined", name),
            EvalErrorType::NumArgs(func, expected, received) => write!(
                f,
                "Incorrect number of args to {}. Expected: {}, Received: {}",
                func, expected, received
            ),
            EvalErrorType::InvalidArgName(func, arg) => {
                write!(f, "{} does not have an arg named {}", func, arg)
            }
            EvalErrorType::MissingArgs(func, args) => {
                write!(f, "Missing args {} for {}", args.join(", "), func)
            }
            EvalErrorType::MissingMain => write!(f, "Missing main shape"),
        }
    }
}

pub type EvalError = Error<EvalErrorType>;
