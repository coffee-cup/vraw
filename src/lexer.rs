use std::iter::Peekable;

use crate::error::*;
use crate::utils::*;

#[derive(Debug, PartialEq, Clone)]
pub enum TokenType {
    LParen,
    RParen,
    LCurly,
    RCurly,
    Times,
    Divide,
    Plus,
    Minus,
    Equals,
    Compare,
    Colon,
    Comma,
    Number(f64),
    Ident(String),
}

#[derive(Debug, PartialEq, Clone)]
pub struct Token {
    token_type: TokenType,
    token_pos: Range,
}

pub struct Lexer<'a> {
    iter: Peekable<std::str::Chars<'a>>,
    line: u32,
    column: u32,
}

pub fn lex(input: &String) -> Result<Vec<Token>, Error> {
    let mut tokens = vec![];
    let mut lexer = Lexer::new(input);

    while let Some(t) = lexer.next()? {
        tokens.push(t);
    }

    Ok(tokens)
}

impl Token {
    pub fn token_type(&self) -> TokenType {
        self.token_type.clone()
    }

    pub fn token_pos(&self) -> Range {
        self.token_pos
    }
}

impl<'a> Lexer<'a> {
    pub fn new(input: &'a String) -> Lexer<'a> {
        Lexer {
            iter: input.trim().chars().peekable(),
            line: 0,
            column: 0,
        }
    }

    fn pos(&self) -> Pos {
        Pos {
            line: self.line,
            column: self.column,
        }
    }

    fn token(&self, token_type: TokenType, start: Pos, end: Pos) -> Option<Token> {
        Some(Token {
            token_type: token_type,
            token_pos: Range {
                start: start,
                end: end,
            },
        })
    }

    fn advance(&mut self, token_type: TokenType) -> Result<Option<Token>, Error> {
        let start = self.pos();

        self.iter.next();
        self.column += 1;

        let end = self.pos();

        let token = self.token(token_type, start, end);
        Ok(token)
    }

    fn advance_whitespace(&mut self) -> bool {
        match self.iter.peek() {
            Some(' ') | Some('\t') => {
                self.column += 1;
                self.iter.next();
                true
            }
            Some('\n') | Some('\r') => {
                self.column = 0;
                self.line += 1;
                self.iter.next();
                true
            }
            _ => false,
        }
    }

    fn forward(&mut self) -> Option<char> {
        self.column += 1;
        self.iter.next()
    }

    fn match_next(&mut self, pred: &Fn(char) -> bool) -> Option<char> {
        let c = match self.iter.peek() {
            Some(c) => *c,
            None => return None,
        };

        if pred(c) {
            self.forward();
            Some(c)
        } else {
            None
        }
    }

    fn consume_ident(&mut self) -> Result<Option<Token>, Error> {
        let start = self.pos();
        let mut id = String::new();

        match self.match_next(&is_alpha) {
            Some(c) => id.push(c),
            None => {
                return Err(Error::new(
                    "identifiers must start with letter.".to_owned(),
                    Some(self.pos()),
                ))
            }
        };

        while let Some(c) = self.match_next(&is_alphanum) {
            id.push(c);
        }
        let end = self.pos();

        let token = self.token(TokenType::Ident(id), start, end);
        Ok(token)
    }

    fn consume_number(&mut self) -> Result<Option<Token>, Error> {
        let start = self.pos();

        let mut str = String::new();
        let mut seen_dot = false;

        while let Some(c) = self.iter.peek() {
            if *c == '.' && !seen_dot {
                seen_dot = true;
                str.push(self.forward().unwrap());
            } else if is_digit(*c) {
                str.push(self.forward().unwrap());
            } else {
                break;
            }
        }

        let end = self.pos();

        let n = str.parse::<f64>().unwrap();

        Ok(self.token(TokenType::Number(n), start, end))
    }

    fn next(&mut self) -> Result<Option<Token>, Error> {
        // ignore whitespace
        while self.advance_whitespace() {}

        let c = match self.iter.peek() {
            Some(c) => c,
            None => return Ok(None),
        };

        let token = match c {
            '(' => self.advance(TokenType::LParen),
            ')' => self.advance(TokenType::RParen),
            '{' => self.advance(TokenType::LCurly),
            '}' => self.advance(TokenType::RCurly),
            '*' => self.advance(TokenType::Times),
            '/' => self.advance(TokenType::Divide),
            '+' => self.advance(TokenType::Plus),
            '-' => self.advance(TokenType::Minus),
            ':' => self.advance(TokenType::Colon),
            ',' => self.advance(TokenType::Comma),
            '=' => {
                let start = self.pos();
                self.forward();
                let end = self.pos();
                match self.iter.peek() {
                    Some('=') => self.advance(TokenType::Compare),
                    _ => Ok(self.token(TokenType::Equals, start, end)),
                }
            }
            'a'...'z' => self.consume_ident(),
            '0'...'9' => self.consume_number(),
            _ => Err(Error::new(
                format!("unexpected character: {:?}", c),
                Some(self.pos()),
            )),
        };

        token
    }
}

fn is_alpha(c: char) -> bool {
    match c {
        'a'...'z' | 'A'...'Z' => true,
        _ => false,
    }
}

fn is_digit(c: char) -> bool {
    match c {
        '0'...'9' => true,
        _ => false,
    }
}

fn is_alphanum(c: char) -> bool {
    is_alpha(c) || is_digit(c)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn lex_identifier() {
        let tokens = lex(&"a1234".to_owned());
        assert_eq!(
            tokens,
            Ok(vec![Token {
                token_type: TokenType::Ident("a1234".to_owned()),
                token_pos: Range {
                    start: Pos { line: 0, column: 0 },
                    end: Pos { line: 0, column: 5 }
                }
            }])
        );
    }

    #[test]
    fn lex_integer() {
        let tokens = lex(&"10".to_owned());
        assert_eq!(
            tokens,
            Ok(vec![Token {
                token_type: TokenType::Number(10.0),
                token_pos: Range {
                    start: Pos { line: 0, column: 0 },
                    end: Pos { line: 0, column: 2 }
                }
            }])
        )
    }

    #[test]
    fn lex_float() {
        let tokens = lex(&"10.123".to_owned());
        assert_eq!(
            tokens,
            Ok(vec![Token {
                token_type: TokenType::Number(10.123),
                token_pos: Range {
                    start: Pos { line: 0, column: 0 },
                    end: Pos { line: 0, column: 6 }
                }
            }])
        )
    }
}
