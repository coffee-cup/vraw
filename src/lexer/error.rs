use std::fmt;

use crate::error::Error;
use crate::utils::*;

#[derive(Debug, PartialEq)]
pub enum LexerErrorType {
    InvalidIdentifier,
    StringNeverTerminated,
    UnexpectedCharacter(char),
}

impl fmt::Display for LexerErrorType {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            LexerErrorType::InvalidIdentifier => write!(f, "Identifier must start with a letter"),
            LexerErrorType::StringNeverTerminated => write!(f, "String is never terminated"),
            LexerErrorType::UnexpectedCharacter(c) => write!(f, "Unexpected character `{}`", c),
        }
    }
}

pub type LexerError = Error<LexerErrorType>;
