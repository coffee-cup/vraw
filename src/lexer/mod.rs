use std::iter::Peekable;

use crate::error::*;
use crate::utils::*;

mod error;
use error::LexerErrorType::*;
use error::*;

type LexerResult<T> = Result<T, LexerError>;

fn lexer_error<T>(error_type: LexerErrorType, pos: Pos) -> LexerResult<T> {
    Err(LexerError::new(error_type, pos))
}

#[derive(Debug, PartialEq, Clone)]
pub enum TokenType {
    LParen,
    RParen,
    LCurly,
    RCurly,
    Times,
    Divide,
    Plus,
    Minus,
    Equals,
    Compare,
    Colon,
    Comma,
    Number(f64),
    Ident(String),
    String(String),
}

#[derive(Debug, PartialEq, Clone)]
pub struct Token {
    token_type: TokenType,
    token_pos: Range,
}

impl HasPos for Token {
    fn pos(&self) -> Pos {
        self.token_pos().start
    }
}

pub struct Lexer<'a> {
    iter: Peekable<std::str::Chars<'a>>,
    line: u32,
    column: u32,
}

pub fn lex(input: &String) -> LexerResult<Vec<Token>> {
    let mut tokens = vec![];
    let mut lexer = Lexer::new(input);

    while let Some(t) = lexer.next()? {
        tokens.push(t);
    }

    Ok(tokens)
}

impl Token {
    pub fn token_type(&self) -> TokenType {
        self.token_type.clone()
    }

    pub fn token_pos(&self) -> Range {
        self.token_pos
    }
}

impl<'a> Lexer<'a> {
    pub fn new(input: &'a String) -> Lexer<'a> {
        Lexer {
            iter: input.trim().chars().peekable(),
            line: 0,
            column: 0,
        }
    }

    fn pos(&self) -> Pos {
        Pos {
            line: self.line,
            column: self.column,
        }
    }

    fn token(&self, token_type: TokenType, start: Pos, end: Pos) -> Option<Token> {
        Some(Token {
            token_type: token_type,
            token_pos: Range {
                start: start,
                end: end,
            },
        })
    }

    fn advance(&mut self, token_type: TokenType) -> LexerResult<Option<Token>> {
        let start = self.pos();

        self.iter.next();
        self.column += 1;

        let end = self.pos();

        let token = self.token(token_type, start, end);
        Ok(token)
    }

    fn advance_whitespace(&mut self) -> bool {
        match self.iter.peek() {
            Some(' ') | Some('\t') => {
                self.column += 1;
                self.iter.next();
                true
            }
            Some('\n') | Some('\r') => {
                self.column = 0;
                self.line += 1;
                self.iter.next();
                true
            }
            _ => false,
        }
    }

    fn forward(&mut self) -> Option<char> {
        self.column += 1;
        self.iter.next()
    }

    fn match_next(&mut self, pred: impl Fn(char) -> bool) -> Option<char> {
        let c = match self.iter.peek() {
            Some(c) => *c,
            None => return None,
        };

        if pred(c) {
            self.forward();
            Some(c)
        } else {
            None
        }
    }

    fn consume_string(&mut self) -> LexerResult<Option<Token>> {
        let start = self.pos();
        let mut s = String::new();

        self.forward();
        loop {
            let c = self.forward();

            match c {
                None => return lexer_error(StringNeverTerminated, start),
                Some('"') => break,
                Some(c) => s.push(c),
            }
        }

        let end = self.pos();

        let token = self.token(TokenType::String(s), start, end);
        Ok(token)
    }

    fn consume_ident(&mut self) -> LexerResult<Option<Token>> {
        let start = self.pos();
        let mut id = String::new();

        match self.match_next(&is_alpha) {
            Some(c) => id.push(c),
            None => return lexer_error(InvalidIdentifier, self.pos()),
        };

        while let Some(c) = self.match_next(&is_alphanum) {
            id.push(c);
        }
        let end = self.pos();

        let token = self.token(TokenType::Ident(id), start, end);
        Ok(token)
    }

    fn consume_number(&mut self) -> LexerResult<Option<Token>> {
        let start = self.pos();

        let mut str = String::new();
        let mut seen_dot = false;

        while let Some(c) = self.iter.peek() {
            if *c == '.' && !seen_dot {
                seen_dot = true;
                str.push(self.forward().unwrap());
            } else if is_digit(*c) {
                str.push(self.forward().unwrap());
            } else {
                break;
            }
        }

        let end = self.pos();

        let n = str.parse::<f64>().unwrap();

        Ok(self.token(TokenType::Number(n), start, end))
    }

    fn next(&mut self) -> LexerResult<Option<Token>> {
        // ignore whitespace
        while self.advance_whitespace() {}

        let c = match self.iter.peek() {
            Some(c) => c,
            None => return Ok(None),
        };

        let token = match c {
            '(' => self.advance(TokenType::LParen),
            ')' => self.advance(TokenType::RParen),
            '{' => self.advance(TokenType::LCurly),
            '}' => self.advance(TokenType::RCurly),
            '*' => self.advance(TokenType::Times),
            '/' => self.advance(TokenType::Divide),
            '+' => self.advance(TokenType::Plus),
            '-' => self.advance(TokenType::Minus),
            ':' => self.advance(TokenType::Colon),
            ',' => self.advance(TokenType::Comma),
            '=' => {
                let start = self.pos();
                self.forward();
                let end = self.pos();
                match self.iter.peek() {
                    Some('=') => self.advance(TokenType::Compare),
                    _ => Ok(self.token(TokenType::Equals, start, end)),
                }
            }
            '"' => self.consume_string(),
            'a'...'z' => self.consume_ident(),
            '0'...'9' => self.consume_number(),
            _ => return lexer_error(UnexpectedCharacter(*c), self.pos()),
        };

        token
    }
}

fn is_alpha(c: char) -> bool {
    match c {
        'a'...'z' | 'A'...'Z' => true,
        _ => false,
    }
}

fn is_digit(c: char) -> bool {
    match c {
        '0'...'9' => true,
        _ => false,
    }
}

fn is_alphanum(c: char) -> bool {
    is_alpha(c) || is_digit(c)
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_debug_snapshot_matches;

    #[test]
    fn lex_identifier() {
        let tokens = lex(&"a1234".to_owned());
        assert_debug_snapshot_matches!(tokens);
    }

    #[test]
    fn lex_integer() {
        let tokens = lex(&"10".to_owned());
        assert_debug_snapshot_matches!(tokens);
    }

    #[test]
    fn lex_float() {
        let tokens = lex(&"10.123".to_owned());
        assert_debug_snapshot_matches!(tokens);
    }

    #[test]
    fn lex_string() {
        let tokens = lex(&"\"hello\"".to_owned());
        assert_debug_snapshot_matches!(tokens);
    }

    #[test]
    fn lex_parens() {
        let tokens = lex(&"()".to_owned());
        assert_debug_snapshot_matches!(tokens);
    }

    #[test]
    fn lex_curlies() {
        let tokens = lex(&"{}".to_owned());
        assert_debug_snapshot_matches!(tokens);
    }

    #[test]
    fn lex_operators() {
        let tokens = lex(&"*/+-===".to_owned());
        assert_debug_snapshot_matches!(tokens);
    }

    #[test]
    fn lex_special() {
        let tokens = lex(&":,".to_owned());
        assert_debug_snapshot_matches!(tokens);
    }
}
