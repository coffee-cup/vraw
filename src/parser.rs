use std::iter::Peekable;
use std::slice::Iter;

use crate::error::*;
use crate::lexer::*;
use crate::utils::*;

type Ident = String;
type Arg = Ident;

#[derive(Debug, PartialEq)]
pub struct Program {
    shapes: Vec<Shape>,
    pos: Pos,
}

#[derive(Debug, PartialEq)]
pub struct Shape {
    name: Ident,
    args: Vec<Arg>,
    stmts: Vec<Stmt>,
    pos: Pos,
}

#[derive(Debug, PartialEq)]
pub struct NamedArg {
    name: Ident,
    expr: Expr,
    pos: Pos,
}

#[derive(Debug, PartialEq)]
pub struct Stmt {
    expr: Expr,
    pos: Pos,
}

#[derive(Debug, PartialEq)]
pub enum Expr {
    Name(Ident, Range),
    FunCall(Ident, Vec<NamedArg>, Range),
    Literal(Literal, Range),
    Binary(Box<Expr>, BinOp, Box<Expr>, Pos),
    Unary(UnOp, Box<Expr>, Pos),
}

#[derive(Debug, PartialEq)]
pub enum Literal {
    Number(f64),
    String(String),
}

#[derive(Debug, PartialEq)]
pub enum BinOp {
    Mul,
    Div,
    Add,
    Sub,
}

#[derive(Debug, PartialEq)]
pub enum UnOp {
    Neg,
}

#[derive(Copy, Clone)]
enum Precedence {
    Conditional = 20,
    Sum = 30,
    Product = 40,
    Exponent = 50,
    Prefix = 60,
    Postfix = 70,
    Call = 80,
}

fn prec(p: Precedence) -> u32 {
    p as u32
}

pub struct Parser<'a> {
    input: Peekable<Iter<'a, Token>>,
}

impl Token {
    // null denotation
    fn nud(&self, parser: &mut Parser) -> Result<Expr, Error> {
        match self.token_type() {
            TokenType::Ident(s) => Ok(Expr::Name(s, self.token_pos())),
            TokenType::Number(n) => Ok(Expr::Literal(Literal::Number(n), self.token_pos())),
            TokenType::Minus => {
                let e = parser.expression(prec(Precedence::Prefix))?;
                Ok(Expr::Unary(UnOp::Neg, Box::new(e), self.token_pos().start))
            }
            TokenType::LParen => {
                let e = parser.expression(0)?;
                if let Some(TokenType::RParen) = parser.input.next().map(|t| t.token_type()) {
                    Ok(e)
                } else {
                    Err(Error::new(
                        "unbalanced paren".to_owned(),
                        Some(self.token_pos().start),
                    ))
                }
            }
            _ => Err(Error::new(
                "expecting literal".to_owned(),
                Some(self.token_pos().start),
            )),
        }
    }

    // left denotation
    fn led(&self, parser: &mut Parser, lhs: Expr) -> Result<Expr, Error> {
        match self.token_type() {
            TokenType::Times | TokenType::Divide | TokenType::Plus | TokenType::Minus => {
                let rhs = parser.expression(self.lbp())?;
                let op = token_to_binop(self.clone()).unwrap();
                Ok(Expr::Binary(
                    Box::new(lhs),
                    op,
                    Box::new(rhs),
                    self.token_pos().start,
                ))
            }
            _ => Err(Error::new(
                "expecting operator".to_owned(),
                Some(self.token_pos().start),
            )),
        }
    }

    // left binding power
    fn lbp(&self) -> u32 {
        match self.token_type() {
            TokenType::LParen => prec(Precedence::Call),
            TokenType::Times | TokenType::Divide => prec(Precedence::Product),
            TokenType::Plus | TokenType::Minus => prec(Precedence::Sum),
            _ => 0,
        }
    }
}

fn token_to_binop(token: Token) -> Result<BinOp, String> {
    match token.token_type() {
        TokenType::Times => Ok(BinOp::Mul),
        TokenType::Divide => Ok(BinOp::Div),
        TokenType::Plus => Ok(BinOp::Add),
        TokenType::Minus => Ok(BinOp::Sub),
        _ => Err("no binop for token".to_owned()),
    }
}

impl<'a> Parser<'a> {
    pub fn new(input: Iter<'a, Token>) -> Parser<'a> {
        Parser {
            input: input.peekable(),
        }
    }

    pub fn expression(&mut self, rbp: u32) -> Result<Expr, Error> {
        let mut left = self.parse_nud()?;
        while self.next_binds_tighter(rbp) {
            left = self.parse_led(left)?;
        }

        Ok(left)
    }

    fn next_binds_tighter(&mut self, rbp: u32) -> bool {
        self.input.peek().map_or(false, |t| t.lbp() > rbp)
    }

    fn parse_nud(&mut self) -> Result<Expr, Error> {
        self.input
            .next()
            .map_or(Err(Error::new("incomplete".to_owned(), None)), |t| {
                t.nud(self)
            })
    }

    fn parse_led(&mut self, expr: Expr) -> Result<Expr, Error> {
        self.input
            .next()
            .map_or(Err(Error::new("incomplete".to_owned(), None)), |t| {
                t.led(self, expr)
            })
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::lexer;
    use crate::utils::*;

    fn parse_expression(input: &String) -> Result<Expr, Error> {
        let tokens = lexer::lex(input)?;
        let mut parser = Parser::new(tokens.iter());
        parser.expression(0)
    }

    #[test]
    fn parse_number_literal() {
        let ast = parse_expression(&"2".to_owned());
        assert_eq!(
            ast,
            Ok(Expr::Literal(
                Literal::Number(2.0),
                create_range(create_pos(0, 0), create_pos(0, 1))
            ))
        );
    }

    #[test]
    fn parse_simple_identifier() {
        let ast = parse_expression(&"hello".to_owned());
        assert_eq!(
            ast,
            Ok(Expr::Name(
                "hello".to_owned(),
                create_range(create_pos(0, 0), create_pos(0, 5))
            )),
        );
    }

    #[test]
    fn parse_simple_binary_expression() {
        let ast = parse_expression(&"10 + 2".to_owned());
        assert_eq!(
            ast,
            Ok(Expr::Binary(
                Box::new(Expr::Literal(
                    Literal::Number(10.0),
                    create_range(create_pos(0, 0), create_pos(0, 2))
                )),
                BinOp::Add,
                Box::new(Expr::Literal(
                    Literal::Number(2.0),
                    create_range(create_pos(0, 5), create_pos(0, 6))
                )),
                create_pos(0, 3)
            ))
        );
    }
}
