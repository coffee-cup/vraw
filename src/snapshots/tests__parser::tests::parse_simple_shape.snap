---
created: "2019-06-25T19:28:00.475370Z"
creator: insta@0.8.1
source: src/parser.rs
expression: ast
---
Ok(
    Shape {
        name: "circle",
        args: [
            "r"
        ],
        stmts: [
            Expr(
                FunCall(
                    "ellipse",
                    [
                        NamedArg {
                            name: "rx",
                            expr: Name(
                                "r",
                                Range {
                                    start: Pos {
                                        line: 1,
                                        column: 14
                                    },
                                    end: Pos {
                                        line: 1,
                                        column: 15
                                    }
                                }
                            )
                        },
                        NamedArg {
                            name: "ry",
                            expr: Name(
                                "r",
                                Range {
                                    start: Pos {
                                        line: 1,
                                        column: 21
                                    },
                                    end: Pos {
                                        line: 1,
                                        column: 22
                                    }
                                }
                            )
                        }
                    ],
                    Range {
                        start: Pos {
                            line: 1,
                            column: 2
                        },
                        end: Pos {
                            line: 1,
                            column: 22
                        }
                    }
                ),
                Pos {
                    line: 1,
                    column: 2
                }
            )
        ],
        range: Range {
            start: Pos {
                line: 0,
                column: 0
            },
            end: Pos {
                line: 2,
                column: 0
            }
        }
    }
)
